specs/advanced-data-layouts/prollytree/spec.md

```diff
@@ -0,0 +1,467 @@
+
+### Re-Calculate
+
+In order to modify the tree (inserting, updating, or removing one or multiple values) a part of the tree needs to be re-calculated. This section describes an internal procedure that is not exposed to the caller.
+The tree is (partially) re-calculated bottom-up.
+
+Starting at the inserted/modified leaf or node referencing a now deleted/removed leaf, successively walk up the tree.If a removed item was the only item in its node: Remove the node from its parent and Check if the hash/address/CID of a new leaf/node splits the parent chunk (split the parent node if yes).
+    Then, check if a removed leaf/node was a 'splitting node' and the nodes need to be merged (only the last node in a chunk can be a splitting node). If splitting criterion holds for the last item of the node and there is the current node is succeeded by another node on the same level: merge the currend node with the succeeding node. Continue walking up the path. If at root and it is being split: Create a new root that links to the freshly split nodes. Return the new root CID.
+
+![](https://i.imgur.com/xE0id0V.png)
+![](https://i.imgur.com/aYqzXZI.png)
+![](https://i.imgur.com/nuvInkH.png)
+![](https://i.imgur.com/zACS8gS.png)
+
+Pay attention to the fact that the boundary algorithm is not related with the node cid, i.e node cid is got after the boundary is generated and the node is saved in blockstore. Maybe in the future we can combine the boundary algorithm and cid. 
+
+### Put/Remove/Update
+
+First search if the tree already contains the value.
+If the key is not contained, create a new leaf for the value, insert the key-value (key-leaf) pair in the correct node and re-calculate the tree.
+If the key is already stored in the tree and the value is the same, the tree is left unchanged.
+If the key exists and the values differ: Either leave the tree or insert the new value and update the tree.
+If removing the key, remove the `key` and the `value` from the list. After making any modifications to the tree, run the algoritm (Re-Calculate)[#Re-Calculate].
+
+## Structure
+
+IPLD Schema(https://github.com/kenlabs/ptree-bs/blob/main/pkg/prolly/tree/schema.ipldsch)
+
+```
+type ProllyNode struct {
+	config &ChunkConfig
+    level Int
+	keys [Bytes]
+	links nullable [&ProllyNode]
+	values nullable [Any]
+} representation tuple
+
+type WeibullThresholdConfig struct {
+	K Float
+	L Float
+} representation tuple
+
+type RollingHashConfig struct {
+	rollingHashWindow Int
+} representation tuple
+
+type PrefixThresholdConfig struct {
+  chunkingFactor Int
+}
+
+type ChunkStrategy enum {
+  | PrefixThreshold
+  | WeibullThreshold
+  | RollingHash
+} representation string
+
+type ChunkConfig struct {
```

Comment:
Would it be better to use union to replace the multi configs?
type ChunkConfig struct{
    chunkStrategy ChunkStrategy
    minChunkSize Int
    maxChunkSize Int
    strategy      strategy
} representation tuple

type strategy union {
    | WeibullThresholdConfig    "weibull"
    | RollingHashConfig         "rollinghash"
    | PrefixThresholdConfig     "prefix"
} representation keyed

type ChunkStrategy enum {
  | PrefixThreshold
  | WeibullThreshold
  | RollingHash
} representation string

Reply:
Yeah, if we use a keyed union we can replace the `ChunkStrategy` enum. This would mean having unique key names per each strategy which might be annoying if we need reuse. 🤔 Let's try it out though.